### PR TITLE
Add ability to blacklist params. also sanitise keys

### PR DESCRIPTION
**Pull request by pascalwhoop**
## Description
- special chars need to be sanitized before sending to MLflow

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Comment by Galileo-Galilei** [comment]
Hi, I definitely understand why you don't want to log secrets and I am not strictly against this change, but I think the main issue at stake here is that they should not be params in the first place. Avoid logging in mlflow is likely the least of your problems : if some secrets are prams, they will be logged to your remote version control forge alongside with configuration. Even if you use it in a local folder, this means these secrets will be required in a non crypted file in production. 

Can you elaborate on why you need to have sensitive data in the params in the first place?

I suggest you use environment variable to retrieve these sensitive data directly in the node, so you have control over what is logged.

**Comment by pascalwhoop** [comment]
@Galileo-Galilei let me get back on the secrets one, I think you have a point although I think kedro works this way (I agree ENV variables injection is better and we do do that but we still pass them through the catalog into the python functions to keep them pure)

On the sanitization you're OK? The error was that we had properties in a data-fabricator that were something like `customer_ids:STRING[]` as we fabricate the real systems data structures. But that failed because MLFLow API doesn't like `[` in a key, while in yaml it's OK.

**Comment by pascalwhoop** [comment]
Clearly an active discussion, spotted your name again here
https://github.com/kedro-org/kedro/wiki/Credentials-in-kedro

@merelcht @Galileo-Galilei given this plugin takes all params and sends off to 3rd party service, but kedro recommends passing credentials through parameters into nodes, wdyt is best here? 

Maybe just use Pydantic Settings and create a `Secrets` singleton?
https://docs.pydantic.dev/latest/concepts/pydantic_settings/#usage

**Comment by pascalwhoop** [comment]
I removed the secrets removal from the parameters as that feels like a separate and much larger problem than the one where the parameters can have special chars in the key in kedro but not on MLflow. I hope this smaller change is less controversial

**Comment by Galileo-Galilei** [comment]
Hello @pascalwhoop definitely ok to sanitize parameters like this, but I'd like to ensure consistency with mlflow. I was pretty sure it has existed at some point in the past, not sure why it's no longer the case. 

This issue https://github.com/mlflow/mlflow/issues/9425 refers to the code with the exact regex mlflow uses under the hood (and explains the rationale behind this) , and we should use the same one. The implementation should import the mlflow function that performs validation so we don't have to mirror the evolution of the behaviour along with mlflow. Could you update that?

**Comment by pascalwhoop** [comment]
done @Galileo-Galilei

File: tests/framework/hooks/test_hook_log_parameters.py [hunk 1]
```diff
@@ -74,6 +75,60 @@ def dummy_catalog():
...
+    mlflow_tracking_uri = (kedro_project / "mlruns").as_uri()
+    mlflow.set_tracking_uri(mlflow_tracking_uri)
+
+    node_inputs = {f"params:{param_name}": "test_value"}
+
+    bootstrap_project(kedro_project)
+    with KedroSession.create(
+        project_path=kedro_project,
+    ) as session:
+        context = session.load_context()
```
Comment (Galileo-Galilei) [inline tests/framework/hooks/test_hook_log_parameters.py:108]: Just a nitpick, but can we just ignore all this? Since you invoke the hook manually, I don't think it's necessary creating a whole session and contexte, unless I miss something?

File: tests/framework/hooks/test_hook_log_parameters.py [hunk 2]
```diff
@@ -74,6 +75,60 @@ def dummy_catalog():
...
+)
+def test_parameter_name_sanitization(
+    kedro_project, dummy_run_params, param_name, expected_name
+):
+    _write_yaml(
+        kedro_project / "conf" / "local" / "mlflow.yml",
+        dict(
+            tracking=dict(params=dict(long_params_strategy="fail")),
+        ),
+    )
```
Comment (Galileo-Galilei) [inline tests/framework/hooks/test_hook_log_parameters.py:97]: Why is this needed?  It should work without mlflow.yml.

File: tests/framework/hooks/test_hook_log_parameters.py [hunk 3]
```diff
@@ -74,6 +75,60 @@ def dummy_catalog():
...
+
+    node_inputs = {f"params:{param_name}": "test_value"}
+
+    bootstrap_project(kedro_project)
+    with KedroSession.create(
+        project_path=kedro_project,
+    ) as session:
+        context = session.load_context()
+        mlflow_node_hook = MlflowHook()
+        mlflow_node_hook.after_context_created(context)
```
Comment (Galileo-Galilei) [inline tests/framework/hooks/test_hook_log_parameters.py:110]: Not sure you need after_context_created since you already set ``mlflow_tracking_uri`` up just above

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [tool.pytest.ini_options]
-addopts = "--cov=kedro_mlflow --cov-report=html tests/"
```
Comment (Galileo-Galilei) [inline pyproject.toml:2]: Why isn't it necessary anymore?

File: kedro_mlflow/framework/hooks/mlflow_hook.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import platform
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/mlflow_hook.py:1]: import platform does not seem used anywhere?

File: tests/framework/hooks/test_hook_log_parameters.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import platform
```
Comment (Galileo-Galilei) [inline tests/framework/hooks/test_hook_log_parameters.py:1]: ``import platform`` does not seem used anywhere?

File: kedro_mlflow/framework/hooks/mlflow_hook.py [hunk 7]
```diff
@@ -447,4 +454,15 @@ def on_pipeline_error(
...
 
+def validate_and_sanitize_param_name(name: str) -> str:
+    # regex taken from MLFlow codebase: https://github.com/mlflow/mlflow/blob/e40e782b6fcab473159e6d4fee85bc0fc10f78fd/mlflow/utils/validation.py#L140C1-L148C44
+    pattern = r"^[/\w.\- :]*$"
+
+    if re.match(pattern, name):
+        return name
+    else:
+        # Replace invalid characters with underscore
+        return re.sub(r"[^/\w.\- :]", "_", name)
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/mlflow_hook.py:465]: can we add a ``logger.warning`` to explain that the parameter is renamed? This might be confusing.

**Review by Galileo-Galilei — COMMENTED** [review]
Hi, just a couple of questions / suggestions, but I'm ok to merge it without any change if you disagree.

File: pyproject.toml (hunk 4 above)
Comment (pascalwhoop) [reply pyproject.toml:2]: ah my bad, I had removed this to be able to call `pytest ..path..to.test` manually, the `tests/` at the end blocked that from being possible

Comment (Galileo-Galilei) [reply pyproject.toml:2]: (just for the record, I guess you're searching ``pytest -k "regex"`` to execute only some tests based on a regex)

File: tests/framework/hooks/test_hook_log_parameters.py (hunk 1 above)
Comment (Galileo-Galilei) [reply tests/framework/hooks/test_hook_log_parameters.py:107]: My bad, this is needed to call the after_context_created hook, which format and stores the entire configuration.

**Comment by Galileo-Galilei** [comment]
Almost good to go, I just don't understand why he test fails with "valid:param" -> "valid:param". I get ``mlflow.exceptions.MlflowException: Invalid value "valid:param" for parameter 'key' supplied: Names may only contain alphanumerics, underscores (_), dashes (-), periods (.), spaces ( ), colon(:)`` so it is likely on mlflow's side, but i'd like to understand.

I'll resume tomorrow, else it's good to go.

**Comment by Galileo-Galilei** [comment]
It's fixed, it was a windows issue as usual :) I'll merge it tomorrow

**Comment by pascalwhoop** [comment]
hey @Galileo-Galilei thanks, I was a bit swamped and failed to respond. Appreciate the merge
